contracts/Ops.sol

```diff
@@ -1,292 +1,291 @@
+            _feeToken,
+            taskId
         );
     }
 
-    /// @notice Cancel a task so that Gelato can no longer execute it
-    /// @param _taskId The hash of the task, can be computed using getTaskId()
-    function cancelTask(bytes32 _taskId) public override {
+    function _cancelTask(address _taskCreator, bytes32 _taskId) private {
         require(
-            taskCreator[_taskId] == msg.sender,
-            "Ops: cancelTask: Sender did not start task yet"
+            taskCreator[_taskId] == _taskCreator,
+            "Ops.cancelTask: Task not found"
         );
 
-        _createdTasks[msg.sender].remove(_taskId);
+        _createdTasks[_taskCreator].remove(_taskId);
         delete taskCreator[_taskId];
         delete execAddresses[_taskId];
+        delete timedTask[_taskId];
 
-        Time memory time = timedTask[_taskId];
-        bool isTimedTask = time.nextExec != 0;
-        if (isTimedTask) delete timedTask[_taskId];
-
-        emit TaskCancelled(_taskId, msg.sender);
+        emit LibEvents.TaskCancelled(_taskId, _taskCreator);
     }
 
-    function _createTask(
+    // solhint-disable function-max-lines
+    function _exec(
+        bytes32 _taskId,
         address _taskCreator,
         address _execAddress,
-        bytes4 _execSelector,
-        address _resolverAddress,
-        bytes calldata _resolverData,
-        address _feeToken
-    ) internal returns (bytes32 taskId) {
-        bool useTaskTreasuryFunds = _feeToken == address(0);
-        bytes32 resolverHash = getResolverHash(_resolverAddress, _resolverData);
-
-        taskId = getTaskId(
-            _taskCreator,
-            _execAddress,
-            _execSelector,
-            useTaskTreasuryFunds,
-            _feeToken,
-            resolverHash
-        );
-
+        bytes memory _execData,
+        LibDataTypes.Module[] memory _modules,
+        uint256 _txFee,
+        address _feeToken,
+        bool _useTaskTreasuryFunds,
+        bool _revertOnFailure
+    ) private {
         require(
-            taskCreator[taskId] == address(0),
-            "Ops: _createTask: Sender already started task"
+            _taskCreator != address(0) && taskCreator[_taskId] == _taskCreator,
+            "Ops.exec: Task not found"
         );
 
-        _createdTasks[_taskCreator].add(taskId);
-        taskCreator[taskId] = _taskCreator;
-        execAddresses[taskId] = _execAddress;
+        if (!_useTaskTreasuryFunds) {
+            fee = _txFee;
+            feeToken = _feeToken;
+        }
 
-        emit TaskCreated(
+        bool success = LibTaskModule.onExecTask(
+            _taskId,
             _taskCreator,
             _execAddress,
-            _execSelector,
-            _resolverAddress,
-            taskId,
-            _resolverData,
-            useTaskTreasuryFunds,
+            _execData,
+            _modules,
+            _revertOnFailure,
+            taskModuleAddresses
+        );
+
+        if (_useTaskTreasuryFunds) {
```

Comment:
Looking ahead to when we integrate the 1balance system, how would we handle this part here @goums?

Reply:
Looking ahead to when we integrate the 1balance system, how would we handle this part here @goums?

Reply:
I guess we could handle that logic in the taskTreasury, by having a flag which the user can set which redirects the call to the off-chain 1balance contract where we then emit the required events. This would however mean that we will have the call over head of Ops => TaskTreasury => 1Balance system

Reply:
yes, we'll simply emit an event instead of updating the balance if the user is using off-chain balance.
I think @brandonchuah said that they want to have generic events that can be emitted by any contract. Not sure about this but it could help reduce the gas overhead of calling an additional contract

Reply:
Yep, we don't have to do TaskTreasury => 1Balance system call. Just emit a specific event which will be tracked by the 1balance backend

Reply:
I think we are planning on emitting the events for the 1Balance system in a single contract, meaning we would need to do a taskTreasury => 1Balance .call to emit the event over there.